### ISSUE TITLE
Bookmark subentries are not possible

### ISSUE THREAD

**Issue by MartinX3** [masneyb/gftp#114]
1. In the bookmark menu I create a folder.
2. On the folder I do a right click and select `New Item`.
3. The item gets created in the root and not as subentry in my selected folder.

As a workaround I'll edit the `/home/${USER}/.gftp/bookmarks` file now

**Comment by wdlkmpx** [comment]
Yes this problem was always there, currently the bookmarks dialog has some refinements compared to the old dialog, but properly reimplementing some things seems a bit too hard, and I get the feeling GtkTreeView will be removed in GTK5, that makes everything pointless

I'll try to implement the proper behavior, but I agree with the Ardour devs
https://discourse.ardour.org/t/gtk-2-has-been-deprecated-is-there-a-roadmap-for-removal/105214

**Comment by MartinX3** [comment]
Hmm, sad.
I hope sticking to gtk2 will not raise security or compatibility problems while updating the OS and its packages over time.

But I wonder why they should remove the TreeView in GTK5. What should replace it? It's a basic UI element existing in many UI frameworks.

**Comment by wdlkmpx** [comment]
Well the gtk3 port is almost finished,  but removing the deprecated gtk2 stuff requires a major rewrite again, adding pure GTK3 may completely break gftp and some features may longer be possible ... the dynamic bookmarks menu.

I've investigated and have come to the conclusion that gftp must become as simple as Gnome Videos in order to survive another major upgrade
https://linuxreviews.org/GNOME_Videos

**Comment by MartinX3** [comment]
That's a good idea.
I'm doing development myself and if I'm doing the frontend part, I heavy realy on buildin browser features as well (fetchAPI, yes/no Dialogue, file chooser) instead of developing it myself.
It doesn't look perfectly nice, since I can't control the UI part, but the code is far less and easier to migrate to newer framework versions.

**Comment by wdlkmpx** [comment]
Gtk is becoming a mobile UI, that's why it becomes more and more restricted, every major version introduces incredibly huge API changes, that's why many gtk apps have died.

People are still trying to figure out how to replace GtkStatusIcon, I wanted to add a tray icon, but alas it was removed in Gtk4.

**Comment by MartinX3** [comment]
The crazy touch UI move.
Unity died because it wanted to be mobile only on computers.
Also windows 8 was a mistake according to users because of their touch UI on computers.
Many people avoided gnome 3 because of their touch design changes in the past.

I wonder why they want to make these mistakes to restrict a mouse keyboard UI for the sake of touch screens and mobile devices.
A mouse and a keybord won't die, even if sci fi movies wants us to believe that.

They need to release seperate UI elements or their users, the developers, will migrate away.

And especially since this is a low level language project, it is hard to migrate or do a switch.
I wonder if Qt is better or a html solution.
Or wxWidgets. The Money Manager Ex uses wxWidgets and has a nice TreeView.

https://github.com/moneymanagerex/moneymanagerex
https://en.m.wikipedia.org/wiki/List_of_platform-independent_GUI_libraries

**Comment by wdlkmpx** [comment]
They reinvent the wheel with every major release. Important widgets are removed, other widgets are replaced with widgets that do the same things, developers must learn a completely different API, this specially hurts C apps. It's a complete waste of time.

They alienated many users and ruined the toolkit. They have enforced their ugly brand and so all apps are doomed to look bad no matter the desktop environment. Qt is miles ahead, a true cross-platform toolkit that doesn't only focus in the strange mobile needs of a single Linux DE and to hell with everything else.

I implemented the expected behavior in be7e09f55247175713b0c293879d78b4e8827cf8

**Comment by MartinX3** [comment]
Thank you very much!

I see, but Qt is also on mobiles.
So they follow a better strategy.

And looking into the commit reminds me why I prefer high level languages
Less code to implement features

**Comment by wdlkmpx** [comment]
I improved the dialog a bit more.

In the Edit Bookmarks dialog, You can see a menubar. Awesome!

GTK3 supports GtkMenu, GtkMenuItem, GtkMenuBar, etc. It doesn't deprecate those widgets, but they removed for GTK4. Yes they were removed.

In GTK3 the recommended way to create menus is through GtkApplication & GMenu, GtkApplication replaces the Gtk Main Loop and will break some apps (i.e. gftp), and GtkApplication only supports 1 menubar, and the same menubar can be displayed in several windows, but I don't want that. *Ugh*.

I figure out that I need to create another GtkApplication just for the Edit Bookmarks dialog, and destroy it after closing the dialog, and I do know how to, after practicing and breaking the law.

But the current stuff works ok with GTK3, of course for GTK4 many widgets are gone, and so a major rewrite begins. The transition to GtkApplication & GMenu, about 1500-2000 lines must be added, and the dynamic bookmarks menu must be reimplemented using even lower level stuff thank GTK2 (GtkUIManager), at this point something might happen, maybe the realization that it's too advanced or not possible.

**Comment by MartinX3** [comment]
https://developer.gnome.org/gtk4/stable/ch41s02.html#id-1.7.4.4.74 
Doing GtkPopoverMenu and GtkPopoverMenuBar help you?
If I look in their upgrade guides the provide at least replacements.

But if 1500-2000 lines of code needs to be added, I would better look for a different UI framework or if some code is double and can be removed.
